test/Generics/conditional_conformances.swift

```diff
@@ -0,0 +1,204 @@
+// RUN: %FileCheck %s < %t.dump
+
+protocol P1 {}
+protocol P2 {}
+protocol P3 {}
+protocol P4: P1 {}
+
+protocol Assoc { associatedtype AT }
+
+func takes_P2<X: P2>(_: X) {}
+// expected-note@-1{{in call to function 'takes_P2'}}
+// expected-note@-2{{in call to function 'takes_P2'}}
+// expected-note@-3{{in call to function 'takes_P2'}}
+// expected-note@-4{{in call to function 'takes_P2'}}
+// expected-note@-5{{in call to function 'takes_P2'}}
+// expected-note@-6{{in call to function 'takes_P2'}}
+// expected-note@-7{{in call to function 'takes_P2'}}
+// expected-note@-8{{in call to function 'takes_P2'}}
+// expected-note@-9{{in call to function 'takes_P2'}}
+// expected-note@-10{{in call to function 'takes_P2'}}
+
+struct Free<T> {}
+// CHECK-LABEL: ExtensionDecl line={{.*}} base=Free<T>
+// CHECK-NEXT: (normal_conformance type=Free<T> protocol=P2
+// CHECK-NEXT:   conforms_to: τ_0_0 P1)
+extension Free: P2 where T: P1 {}
+func free_good<U: P1>(_: U) {
+    takes_P2(Free<U>())
+}
+func free_bad<U>(_: U) {
+    takes_P2(Free<U>())
+    // expected-error@-1{{generic parameter 'X' could not be inferred}}
+}
+
+struct Constrained<T: P1> {}
+// CHECK-LABEL: ExtensionDecl line={{.*}} base=Constrained<T>
+// CHECK-NEXT: (normal_conformance type=Constrained<T> protocol=P2
+// CHECK-NEXT:   conforms_to: τ_0_0 P3)
+extension Constrained: P2 where T: P3 {}
+func constrained_good<U: P1 & P3>(_: U) {
+    takes_P2(Constrained<U>())
+}
+func constrained_bad<U: P1>(_: U) {
+    takes_P2(Constrained<U>())
+    // expected-error@-1{{generic parameter 'X' could not be inferred}}
+}
+
+struct RedundantSame<T: P1> {}
+// CHECK-LABEL: ExtensionDecl line={{.*}} base=RedundantSame<T>
+// CHECK-NEXT: (normal_conformance type=RedundantSame<T> protocol=P2)
+extension RedundantSame: P2 where T: P1 {}
+// expected-warning@-1{{redundant conformance constraint 'T': 'P1'}}
+
+struct RedundantSuper<T: P4> {}
+// CHECK-LABEL: ExtensionDecl line={{.*}} base=RedundantSuper<T>
+// CHECK-NEXT: (normal_conformance type=RedundantSuper<T> protocol=P2)
+extension RedundantSuper: P2 where T: P1 {}
+// expected-warning@-1{{redundant conformance constraint 'T': 'P1'}}
+
+struct OverlappingSub<T: P1> {}
+// CHECK-LABEL: ExtensionDecl line={{.*}} base=OverlappingSub<T>
+// CHECK-NEXT: (normal_conformance type=OverlappingSub<T> protocol=P2
+// CHECK-NEXT:   conforms_to: τ_0_0 P4)
+extension OverlappingSub: P2 where T: P4 {}
+func overlapping_sub_good<U: P4>(_: U) {
+    takes_P2(OverlappingSub<U>())
+}
+func overlapping_sub_bad<U: P1>(_: U) {
+    takes_P2(OverlappingSub<U>())
+    // expected-error@-1{{generic parameter 'X' could not be inferred}}
+}
+
+
+struct SameType<T> {}
+// CHECK-LABEL: ExtensionDecl line={{.*}} base=SameType<Int>
+// CHECK-NEXT: (normal_conformance type=SameType<Int> protocol=P2
+// CHECK-NEXT:   same_type: τ_0_0 Int)
+extension SameType: P2 where T == Int {}
+// FIXME: the compiler gets this... exactly backwards. :( For the incorrectly
```

Comment:
@DougGregor This test is fun. It looks like with the `T == Int` the `SameType<T>` is eagerly "resolved" to `SameType<Int>` in conformances, meaning `SameType<Int>` isn't treated as a specialised conformance, and that's generally confusing the compiler.